src/formatting/reorder.rs

```diff
@@ -228,14 +228,20 @@ fn rewrite_reorderable_or_regroupable_items(
             for (item, list_item) in normalized_items.iter_mut().zip(list_items) {
                 item.list_item = Some(list_item.clone());
             }
-            match context.config.imports_granularity() {
-                ImportGranularity::Crate => {
-                    normalized_items = merge_use_trees(normalized_items, SharedPrefix::Crate)
-                }
+            normalized_items = match context.config.imports_granularity() {
+                ImportGranularity::Crate => merge_use_trees(normalized_items, SharedPrefix::Crate),
                 ImportGranularity::Module => {
-                    normalized_items = merge_use_trees(normalized_items, SharedPrefix::Module)
+                    merge_use_trees(normalized_items, SharedPrefix::Module)
+                }
+                ImportGranularity::Item => {
+                    merge_use_trees(normalized_items, SharedPrefix::NoPrefix)
+                }
+                ImportGranularity::Preserve => normalized_items,
+            };
+            for item in normalized_items.iter_mut() {
```

Comment:
Should this transformation be part of merge_use_trees?

Reply:
(Although this transformation should be skipped altogether, IIUC)

Reply:
I moved the logic to merge_use_trees, and changed it to rewrite paths ending in `::self` into `::{self}`. This also allowed me to remove some logic from the start of `UseTree::flatten`, which did the same thing but only for one-element lists.

Perhaps this should be part of `UseTree::normalize`, and that should be called again post-merge? Also, I noticed it seems to remove final `::self`: https://github.com/rust-lang/rustfmt/blob/a97fd770e322714933f6f05b6244303c50199dd3/src/formatting/imports.rs#L430